lib/src/rules/avoid_dynamic_calls.dart

```diff
@@ -0,0 +1,245 @@
+    } else {
+      return false;
+    }
+  }
+
+  void _lintIfDynamicOrFunction(Expression node) {
+    final staticType = node?.staticType;
+    if (staticType == null) {
+      return;
+    }
+    if (staticType.isDynamic) {
+      rule.reportLint(node);
+    }
+    if (staticType.isDartCoreFunction) {
+      rule.reportLint(node);
+    }
+  }
+
+  @override
+  void visitAssignmentExpression(AssignmentExpression node) {
+    if (node.readType?.isDynamic != true) {
+      // An assignment expression can only be a dynamic call if it is a
+      // "compound assignment" (i.e. such as `x += 1`); so if `readType` is not
+      // dynamic, we don't need to check further.
+      return;
+    }
+    if (node.operator.type == TokenType.QUESTION_QUESTION_EQ) {
+      // x ??= foo is not a dynamic call.
+      return;
+    }
+    rule.reportLint(node);
+  }
+
+  @override
+  void visitBinaryExpression(BinaryExpression node) {
+    if (!node.operator.isUserDefinableOperator) {
+      // As of 2021-01-13, the only non-virtual operator that has impact on
+      // dynamic calls is "!=", as "a != b" is processed something like
+      // !(a == b), invoking operator== on a (virtual-call).
+      if (node.operator.type != TokenType.BANG_EQ) {
+        return;
+      }
+    }
+    switch (node.operator.type) {
+      case TokenType.EQ_EQ:
+      case TokenType.BANG_EQ:
+        if (node.rightOperand is NullLiteral) {
+          // == and != are special-cased in the dart language to be guaranteed
+          // non-virtual calls iff the RHS is a "null" literal. This allows fast
+          // comparisons, such as "a == null", without invoking a user-defined
+          // operator.
+          return;
+        }
+        break;
+    }
+    _lintIfDynamic(node.leftOperand);
+    // We don't check node.rightOperand, because that is an implicit cast, not a
+    // dynamic call (the call itself is based on leftOperand). While it would be
+    // useful to do so, it is better solved by other more specific lints to
+    // disallow implicit casts from dynamic.
+  }
+
+  @override
+  void visitFunctionExpressionInvocation(FunctionExpressionInvocation node) {
+    if (node.function != null) {
+      _lintIfDynamicOrFunction(node.function);
+    }
+  }
+
+  @override
+  void visitIndexExpression(IndexExpression node) {
+    _lintIfDynamic(node.realTarget);
+  }
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    final receiverWasDynamic = _lintIfDynamic(node.realTarget);
+    if (!receiverWasDynamic && node.target == null) {
+      _lintIfDynamicOrFunction(node.function);
+    }
+  }
+
+  @override
+  void visitPrefixExpression(PrefixExpression node) {
+    if (_lintIfDynamic(node.operand)) {
+      return;
+    }
+    switch (node.operator.type) {
+      case TokenType.PLUS_PLUS:
+      case TokenType.MINUS_MINUS:
+        // The ++ and -- operator expressions don't resolve a static type.
+        if (node.staticType?.isDynamic == true) {
```

Comment:
Could you clarify what kind of a code we are trying to check here?

Reply:
Cases where we're applying `++` or `--` to something with type `dynamic`, such as
```
void f(dynamic x) {
  x++;
}
```

Reply:
I clarified it @scheglov!